qtgui/SourceImage.cpp

```diff
@@ -149,49 +163,91 @@ int SourceImage::clipY() const
+{
+  if (newStartingPoint.x() >= 0)
+  {
+    return pointsToClippingInfo(newClipTopLeft, newClipBottomRight);
+  }
+  return pointsToClippingInfo(clipTopLeft, clipBottomRight);
+}
+
 void SourceImage::normalizeLocations()
 {
   double definedAspectRatio = (1. * resultSize.y()) / resultSize.x();
-  int tempHeight = newBottomRight.y() - newTopLeft.y();
-  int tempWidth = newBottomRight.x() - newTopLeft.x();
-  double temporaryAspectRatio = (1. * tempHeight) / tempWidth;
-  int height = bottomRight.y() - topLeft.y();
-  int width = bottomRight.x() - topLeft.x();
-  double actualAspectRatio = (1. * height) / width;
-
-  if (std::abs(definedAspectRatio - temporaryAspectRatio) > aspectRatioMaxDelta)
+
+  QRectF bounds = boundingRect();
+
+  adjustToAspectRatio(topLeft, bottomRight, definedAspectRatio, bounds);
+  adjustToAspectRatio(newTopLeft, newBottomRight, definedAspectRatio, bounds);
+
+  clipTopLeft.setX(topLeft.x() * image.width() / bounds.width());
+  clipTopLeft.setY(topLeft.y() * image.height() / bounds.height());
+  clipBottomRight.setX(bottomRight.x() * image.width() / bounds.width());
+  clipBottomRight.setY(bottomRight.y() * image.height() / bounds.height());
+
+  newClipTopLeft.setX(newTopLeft.x() * image.width() / bounds.width());
+  newClipTopLeft.setY(newTopLeft.y() * image.height() / bounds.height());
+  newClipBottomRight.setX(newBottomRight.x() * image.width() / bounds.width());
+  newClipBottomRight.setY(newBottomRight.y() * image.height() / bounds.height());
+
+  newClipping();
+}
+
+namespace
+{
+  qreal boundedMin(qreal val1, qreal val2, qreal boundary)
   {
-    // adjust the smaller one
-    if (tempWidth > tempHeight)
-    {
-      newBottomRight.setY(newTopLeft.y() + tempWidth * definedAspectRatio);
-      tempWidth = newBottomRight.x() - newTopLeft.x();
-    }
-    else
+    return std::max(std::min(val1, val2), boundary);
+  }
+  qreal boundedMax(qreal val1, qreal val2, qreal boundary)
+  {
+    return std::min(std::max(val1, val2), boundary);
+  }
+  void adjustToAspectRatio(const QPointF& topLeft, QPointF& bottomRight, double targetAspectRatio, const QRectF& bounds)
+  {
+    double actualHeight = bottomRight.y() - topLeft.y();
+    double actualWidth = bottomRight.x() - topLeft.x();
+    double actualAspectRatio =  actualHeight / actualWidth;
+    static const double aspectRatioMaxDelta{ 0.02 };
+
+    if (std::abs(targetAspectRatio - actualAspectRatio) < aspectRatioMaxDelta)
     {
-      newBottomRight.setX(newTopLeft.x() + tempHeight / definedAspectRatio);
-      tempHeight = newBottomRight.y() - newTopLeft.y();
+      return;
     }
-  }
 
-  if (std::abs(definedAspectRatio - actualAspectRatio) > aspectRatioMaxDelta)
-  {
     // adjust the smaller one
-    if (width > height)
+    if (actualWidth > actualHeight)
     {
-      bottomRight.setY(topLeft.y() + width * definedAspectRatio);
-      height = bottomRight.y() - topLeft.y();
+      qreal newBottomRightY{ topLeft.y() + actualWidth * targetAspectRatio };
```

Comment:
it would be so great to extract this further. however, a per-coordinate implementation requires seven arguments (or a dedicated argument type that then doesn't necessarily forward a reference to the `QPointF` instances on the `SourceImage` - decided not to be worth it.